File: uva/459.py
from sys import stdin, stdout

def rdl():
	return stdin.readline().replace('\n','')

def rdi():
	return int(rdl())

def rds():
	return rdl.spli(' ')

def find(index, sets):
	if (index == sets[index]):
		return index
	sets[index] = find(sets[index], sets)
	return sets[index]

def union_find(u, v, sets, num_sets):
	find_u = find(u, sets)
	find_v = find(v, sets)
	if (find_u != find_v):
		sets[find_v] = find_u
		return num_sets - 1 # reduce to num_sets
	return num_sets # do not reduce to num_sets

def toInt(character):
	return ord(character) - ord('A')

def main():
	tests = rdi()
	rdl() # Read blank
	while(tests):
		max_letter = rdl()
		max_letter = toInt(max_letter) # change read letter to int cero based
		sets = [x for x in range(max_letter + 1)]
		num_sets = max_letter + 1
		line = rdl()
		while(len(line)): # read until blank new line in test case
			first, second = toInt(line[0]), toInt(line[1]) # set first and second to cero based integer from read line
			num_sets = union_find(first, second, sets, num_sets)
			line = rdl()
		stdout.write('{}\n\n'.format(num_sets))
		tests -= 1

if __name__ == '__main__':
	main()

File: uva/459.cpp
#include <iostream>
#include <cstring>

using namespace std;

int num_sets;
int sets[100];

int _find(int index) {
	if (index == sets[index])
		return index;
	return (sets[index] = _find(sets[index]));
}

void union_find(int u, int v) {
	int find_1 = _find(u);
	int find_2 = _find(v);
	cout<<"union_find";
	cout<<find_1<<" "<<find_2<<endl;
	cout<<"union_find"<<endl;
	if (find_1 != find_2) {
		num_sets--;
		sets[find_2] = find_1;
	}
}

int main() {
	int tests;
	string blank, line;
	cin>>tests;
	getline(cin, blank);
	getline(cin, blank);
	while(tests) {
		string max_letter;
		int max_int;
		getline(cin,max_letter);
		max_int = max_letter[0] - 'A';
		num_sets = max_int + 1;
		while(getline(cin, line)) {
			int first = line[0] - 'A';
			int second = line[1] - 'A';
			union_find(first, second);
		}
		tests--;
		if (tests > 0) {
			cout<<num_sets<<endl<<endl;
		} else {
			cout<<num_sets<<endl;
		}
		memset(sets, 0, sizeof(sets));
	}
	return 0;
}